### ISSUE TITLE
Change provided static contracts to return the value instead of `[String: Any]`

### ISSUE THREAD

**Issue by Ybrin** [Boilertalk/Web3.swift#54]
I am talking about `ERC20Contract` in this issue but it applies to all our static contracts.

### Problem

Currently all functions we provide in `ERC20Contract` return a `SolidityInvocation`. This makes our protocols very flexible but also has the disadvantage that we have to make something like that to get the return values we want from the functions:

```Swift
let contract = ... // Some ERC20 contract

firstly {
    contract.balanceOf(address: address).call()
}.done { values in
    guard let balance = values["_balance"] as? BigUInt else {
        // Throw or something...
    }

    // Use balance
    print(balance)
}.catch { error in
    print(error)
}
```

It is very similar with other functions. My point is that we always have to extract our return value from the dictionary `[String: Any]` although we already know what value will be inside it.

### Solution

I suggest we add another layer of abstraction to our static contracts and make our API easier to use by not returning the dictionary but calling a completion with the actual value once we have it.

This means we would automatically call the contract functions instead of giving the user the choice to call them whenever they want. If we want to still provide this option we could add functions like `balanceOfInvocation` to get the `SolidityInvocation` and manage it manually.

### Implementation

Our function signatures in the static contract protocols would look a little bit like that:

```Swift
func balanceOf(address: EthereumAddress, completion: (_ balance: BigUInt?, _ error: Error?))
```

And our PromiseKit extension would look like that:

```Swift
func balanceOf(address: EthereumAddress) -> Promise<BigUInt> {
    return Promise { seal in
        self.balanceOf(address: address, completion: seal.resolve)
    }
}
```

So we could use them like in the following example:

```Swift
let contract = ... // Some ERC20 contract

firstly {
    contract.balanceOf()
}.done { balance in
    // Use balance
    print(balance)
}.catch { error in
    print(error)
}
```

**Comment by Ybrin** [comment]
@pixelmatrix What do you think? I am not sure about how we could make the same for transactions ( the examples are only relevant to read calls). Maybe requiring the private key as another parameter and signing them automatically...?

**Comment by pixelmatrix** [comment]
Certainly seems like a cleaner API. The reason it returns the dictionary currently is just based on the implementation from web3.js, but I agree in the case of swift it makes things more cumbersome. A previous iteration just returned [Any] with only the values, but it seemed like the extra clarity might be helpful in some cases.

It might be challenging to create a standard for how contracts should be written while also specifying types. Perhaps by making an invocation with an associated type or generic values. It's difficult because these protocols are shared with the dynamic contracts as well and those can't easily be typed.

One other thing to consider is just doing this as a convenient utility for the contract templates (ERC20 and ERC721) rather than building it as a standard.

I also think for now this probably only makes sense for calls, since transactions don't return values directly.

**Comment by Ybrin** [comment]
@pixelmatrix Yes I agree, if we want to build a standard for it we will have to try it in real world examples first. Maybe it would be the best to try it out with our templates and build the standard on top of it moving forward...
